// @ledgerhq/cryptoassets/lib/crypto-assets-importer/importers/eip712/eip712.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const _1 = require(".");
const fs_1 = __importDefault(require("fs"));
const eip712Tokens = {
    "1:address:signature": {
        contractName: {
            label: "DeGate Withdrawal",
            signature: "singatureDegate",
        },
        fields: [
            {
                label: "Owner",
                path: "owner",
                signature: "signatureOwner",
            },
        ],
    },
};
const mockedAxios = jest.spyOn(axios_1.default, "get");
describe("import EIP 712 tokens", () => {
    beforeEach(() => {
        mockedAxios.mockImplementation(() => Promise.resolve({ data: eip712Tokens, headers: { etag: "etagHash" } }));
    });
    afterEach(() => {
        jest.clearAllMocks();
    });
    it("should output the file in the correct format", () => __awaiter(void 0, void 0, void 0, function* () {
        const expectedFile = `import EIP712 from "./eip712.json";
export { default as hash } from "./eip712-hash.json";
export default EIP712;
`;
        const mockedFs = (fs_1.default.writeFileSync = jest.fn());
        yield (0, _1.importEIP712)(".");
        expect(mockedAxios).toHaveBeenCalledWith(expect.stringMatching(/.*\/eip712.json/));
        expect(mockedFs).toHaveBeenNthCalledWith(1, "eip712.json", JSON.stringify(eip712Tokens));
        expect(mockedFs).toHaveBeenNthCalledWith(2, "eip712-hash.json", JSON.stringify("etagHash"));
        expect(mockedFs).toHaveBeenNthCalledWith(3, "eip712.ts", expectedFile);
    }));
});
